src/hazelcore/Monads.re

```diff
@@ -6,9 +6,38 @@
 
    In any case, that's a good reference. */
 
-module type MONAD = {
-  [@deriving sexp]
+module type MONAD_BASIC = {
   type t('a);
   let return: 'a => t('a);
   let bind: (t('a), 'a => t('b)) => t('b);
+  let map: [ | `Define_using_bind | `Custom((t('a), 'a => 'b) => t('b))];
```

Comment:
why not use an either type or define a datatype here? hesitant to start using polymorphic variants because it will make it trickier to self-host hazelcore...

Reply:
good point about self-hosting, I'll fix it -- but seriously, why don't more languages have anonymous sums if they have anonymous products.